### PR TITLE
fix(search): return the exact match card search was silently dropping

### LINKED ISSUE #113: bug: the best search result is not returned
### Package version

1.1.0

### Node.js version

21

### Reproduction

```typescript
import { search} from "@mradex77/google-play-scraper";

console.log("== SEARCHING FOR BIEDRONKA");
console.log(await search({ term: "biedronka", num: 2 }));




== SEARCHING FOR biedronka
[
  {
    title: 'Hello JM',
    appId: 'com.servicenow.public.hellojm.requestor',
    url: 'https://play.google.com/store/apps/details?id=com.servicenow.public.hellojm.requestor',
    icon: 'https://play-lh.googleusercontent.com/8fOTzrH07DHrZR87h1t1uau4qqFjv7lOtJvT09E0TgoezVcI4iKb8UXTyxZphw-PQyVUl2zzZpVk1NVBWPUQINc',
    developer: 'JERONIMO MARTINS SGPS, S.A.',
    currency: 'USD',
    price: 0,
    free: true,
    summary: 'Welcome to Hello JM!<br>The app for all the tools and information you need.<br>Created for all employees, Hello JM brings together a community of more than 140,000 employees.<br>Hello JM gives you greater autonomy, agility and efficiency, so you can get everything done more quickly.<br>Everything you need, in the palm of your hand. Anytime, anywhere:<br>•\tView your payslip or request a statement.<br>•\tUpdate your personal details or apply for internal opportunities.<br>•\tStay up to date with company news and campaigns, or access our social programmes.<br>Your day-to-day, faster, smarter and more secure.<br>Enable notifications and always receive the right information at the right time.<br>Hello JM. Connecting everyone, because together we are Jerónimo Martins.<br>Designed for employees and partners of the Jerónimo Martins Group.',
    scoreText: '1.8',
    score: 1.8148148
  },
  {
    title: 'żappka: zakupy, promocje Żabka',
    appId: 'pl.zabka.apb2c',
    url: 'https://play.google.com/store/apps/details?id=pl.zabka.apb2c',
    icon: 'https://play-lh.googleusercontent.com/yCV_VzD4bOWARawCtnkwushzz3SHKwwo8QAGGaPxGogkDzjMHRbr9WpIepM6vYsO2X7gQlDrMSOFDBUJUcmlcQ',
    developer: 'Żabka Polska sp. z o.o.',
    currency: 'USD',
    price: 0,
    free: true,
    summary: 'Zainstaluj aplikację Żappka na swoim smartfonie. Zbieraj żappsy 🐸 i wymieniaj je na nagrody 🎁, podejmuj wyzwania, sprawdzaj aktualne promocje i odkrywaj oferty specjalne w sklepach. Zakupy nigdy nie były tak proste i pełne korzyści!<br><br>Aplikacja Żappka to:<br><br>🎁 Żappsy i nagrody<br><br>Skanuj swój kod podczas zakupów i zbieraj żappsy. Im więcej ku   pujesz, tym więcej zyskujesz! Wymieniaj żappsy na nagrody – baton, gorąca kawa, zimny napój czy ulubiona przekąska. Każdy znajdzie coś dla siebie! Pamiętaj, że żappsy zebrane od 01.11 do 31.10 przepadają, jeśli nie wymienisz ich na nagrody do 31.12. Nie przegap okazji na jeszcze więcej korzyści!<br><br>🛒 Promocje dla Ciebie<br><br>Żappka oferuje promocje skrojone na Twoją  miarę. Odkrywaj oferty specjalne w sklepach i korzystaj z rabatów na ulubione produkty. Każda promocja to okazja do oszczędzania – nawet do 50%! Spraw, by codzienne zakupy były jeszcze bardziej opłacalne. Dzięki aplikacji zawsze jesteś na bieżąco z najnowszymi promocjami i możesz szybko sprawdzić, co jest dostępne w pobliskich sklepach.<br><br>💫 Wyzwania i konkursy<br><b r>Podejmuj wyzwania zakupowe 💪 i zdobywaj niesamowite nagrody – od rejsów 🌴, przez bilety na koncerty, aż po gorące hot dogi 🌭. Bierz udział w konkursach, odkrywaj nowe promocje i zdobywaj dodatkowe żappsy lub produkty. Dzięki aplikacji możesz uczestniczyć w wyjątkowych akcjach i zdobywać bonusy za swoje zakupy – nigdy nie było to tak proste!<br><br>📱 Nowy, intuicyj    ny układ aplikacji<br><br>Odkryj kafelki i poziome menu na dole ekranu, które sprawiają, że poruszanie się po aplikacji jest jeszcze wygodniejsze. Eksploruj Żappkę w nowym wydaniu, odkrywaj szeroki asortyment 🍲 sklepów i korzystaj z najlepszych ofert! Sprawdź wyzwanie powitalne, które przeprowadzi Cię przez wszystkie funkcje nowej aplikacji 🌟 – na końcu czeka nagroda!<b  r><br>🛒 Żabki Nano<br><br>Dodaj swoją kartę i korzystaj ze swobody bezobsługowych zakupów. Odkrywaj Nano w Żappce, sprawdzaj promki oraz produkty za żappsy. Szybko, wygodnie, bezobsługowo – zakupy w promocji jeszcze nigdy nie były tak łatwe! Wszystko, co potrzebujesz, jest teraz w zasięgu ręki.<br><br>🎁 Wygodne usługi w Żabce<br><br>Korzystaj z aplikacji, aby upraszczać   swoje codzienne zakupy. Dzięki integracji z usługami Maczfit, Dietly, Jush i delio, możesz:<br><br>🛒 Zamawiać zakupy z dostawą – szybko i wygodnie.<br>🍲 Wybierać diety z dostawą dostosowane do Twoich potrzeb.<br>🍽️ Korzystać z cateringu dietetycznego.<br><br>To nie wszystko! W aplikacji znajdziesz także:<br><br>📚 Przewodnik po cateringach, aby ułatwić wybór najlepsz    ego rozwiązania.<br>📲 Usługi naprawy telefonu.<br>🎮 Doładowania gier online – idealne dla graczy.<br>🎲 Gry Lotto z szybkimi wynikami.<br>💳 Paysafecard do bezpiecznych płatności.<br>🚚 Paczki – nadawaj i odbieraj bez problemu.<br>📲 Startery GSM dostępne na miejscu.<br><br>Odkrywaj oferty specjalne, korzystaj z szerokiego asortymentu i realizuj codzienne zakupy w s      klepach Żabki. Promocje nigdy nie były tak dostępne, a Twoje codzienne życie jeszcze prostsze i wygodniejsze.<br><br>⚠️ Ważna informacja!<br><br>Numer telefonu jest wymagany, aby korzystać z pełni możliwości aplikacji. Dlaczego?<br><br>Personalizacja: Spersonalizowane oferty i promocje dopasowane do Twoich potrzeb.<br>Funkcjonalności: Uzyskaj dostęp do Żabki Nano i pakietó w.<br>Komunikacja: Otrzymuj powiadomienia o promocjach i ważnych zmianach.<br>Bezpieczeństwo: Zapewnij sobie dodatkową ochronę konta i transakcji.<br>Rozwój: Korzystaj z nowych funkcji i rozszerzeń aplikacji, które stale udoskonalamy.<br><br>Wejdź do aplikacji, wprowadź i potwierdź numer telefonu, aby korzystać ze wszystkich możliwości Żappki! Masz pytania? Napisz na kontakt@zabka.pl. Odkrywaj świat Żappki i ciesz się zakupami jak nigdy dotąd 😍!',
    scoreText: '4.7',
    score: 4.72
  }
]
```

### Expected behavior

I'd expect it to find the first result from the search: https://play.google.com/store/search?q=pl.jmpolska.clos0.mojabiedronka&c=apps

### Actual behavior

It totally omits that most first search result

### PR DESCRIPTION
**Pull request by MrAdex77**
## Summary

**What.** `search()` silently dropped the Play Store exact match card, the large highlighted result Google renders above the ordinary result list. The card was read from one pinned absolute path, `[0, 1, 0, 23]`, which hard codes "slot 23 of the first section". Whenever Google reordered its sections, moved the card off index 23, or drifted the card offer node, the lookup found nothing, the call still succeeded, and the caller got a valid list with its best result missing. Fixes #113.

**Reproduced live, then pinned offline.** Measured on 2026-09-11 for the reporter's own query in the reporter's own storefront (`hl=pl`, `gl=pl`), Google serves three sections for `biedronka`: an unrelated section at index 0 holding an explicit `null` at slot 23, the card at `sections[1][23]`, and the result list at `sections[2][22]`. The pinned path read the `null` and gave up.

| | first results for `biedronka` in the Polish storefront |
| --- | --- |
| `main` | `pl.zabka.apb2c`, `com.gazetki.gazetki`, `pl.com.rossmann.centauros` |
| this branch | **`pl.jmpolska.clos0.mojabiedronka`**, `pl.zabka.apb2c`, `com.gazetki.gazetki` |

That exact page is now recorded as `test/fixtures/search/biedronka-pl.html`, so the reported bug has an offline reproduction that fails against `main`.

**Why it was invisible.** The card is resolved on a separate path from the result list, and every failure on that path is swallowed by `parseOptionalSection`. A missing node emits nothing at all. No unit test and no live test asserted the card was ever surfaced, so the daily scheduled run stayed green throughout.

**How.**

1. `feat(core)`: add the `section-anchor-fallback` integrity reason.
2. `fix(search)`: resolve the card by walking every section instead of reading one pinned path. Tier 1 takes slot 23 of any section, tier 2 falls back to a structural scan one level into each section and reports where it found the card, tier 3 reports an unusable node so a malformed card still raises `optional-section-parse` rather than disappearing.
3. `fix(search)`: declare the card's mirrored `[16][2]` paths as fallbacks for `url`, `price`, `currency` and `free`, so an offer node drift no longer deletes the whole result.
4. `fix(search)`: return the card on a page that carries no result list at all.
5. `fix(search)`: accept only a card that actually extracts, so a node carrying the appId spine and nothing else cannot shadow a real card in a later section.
6. `test` and `docs`: the live tripwire, the recorded multi section fixture, iterator coverage, and the runbook and field audit entries.

**Two corrections to the plan, both backed by measurement.**

The plan anchored the live tripwire on a single owned package id, `com.adex77.WhereAmI`, on the premise that searching a package id forces Google to serve a card. That premise is false. Measured on 2026-09-11, a package id search for that listing serves one section holding a list and no card, in both the `us` and `pl` storefronts, and the same holds for `pl.zabka.apb2c`. Running the planned test verbatim against live Google fails, because Vitest soft assertions still fail the test. Worse, its primary assertion would have passed for the wrong reason: the owned app ranks first in the ordinary list, so the test would have stayed green whether or not the card path worked. The tripwire is therefore anchored on `EXACT_MATCH_CARD_CANDIDATES`, a maintained pool of four package ids Google does serve a card for, under runbook rule 6. It uses `developerId` as the card fingerprint, since only `exactMatchSpecs` produces that field, and it fails only when every anchor has lost its card.

The plan's tier 1 accepted any node whose appId path held a string, which is weaker than what extraction requires. A node carrying only the appId spine at an earlier anchor shadowed a valid card in a later section, and the tier 3 accumulator walked every section rather than the first, adding `optional-section-parse` events on card-less pages where `main` emitted none. Both were confirmed by measurement and both are fixed, with tests that fail against the intermediate implementation.

**Edge cases.** All 17 rows of the design's edge case matrix are covered, plus nine more cases the review surfaced: a decoy card at an earlier anchor, a later anchored card beating an earlier off-anchor one, a non card node parked at a later anchor, a null anchor, several unusable anchors, an unextractable card found only by scanning, cross-node offer mixing, and two iterator paths. The structural check was swept against seven live queries in five storefronts, including pages with sections at slots 21, 22, 24, 25 and 31, with zero false positives.

**Blast radius.** `searchIterator()` shares `fetchSearchFirstPage`, so it gains the same fix with no edit, now covered by its own tests. Nothing else changes: `list`, `similar`, `developer` and the cluster pagination path are untouched. One `IntegrityReason` member is added, nothing is removed or renamed, and no result can be lost, only regained.

**Release.** Minor, driven by the `feat` commit that adds the integrity reason. `IntegrityReason` is a widening union, so an exhaustive switch over it needs a new arm; the README release table now says so. Version and changelog untouched, Release Please owns both.

**Verification.** `pnpm test:e2e` was run against live Google, not only collected with `GP_E2E=0`. All 141 live tests pass, including the new tripwire.

| Gate | Result |
| --- | --- |
| `pnpm lint` | clean |
| `pnpm typecheck` | clean |
| `pnpm format:check` | clean |
| `pnpm test` | 657 passing across 47 files, up from 626 |
| `pnpm test:coverage` | 99.52 statements, 96.47 branches, 99.71 functions, 99.58 lines, all above threshold |
| `pnpm build` | succeeds, emits esm, cjs and types |
| `pnpm check:package` | clean |
| `pnpm docs:api` | clean |
| `pnpm test:e2e` | 141 passing across 18 files, live |

Branch coverage moves up from 96.40 to 96.47 against a floor of 96. One line stays uncovered in `search.ts`, the guard for a sections root that is not an array, which is unreachable through the public API because `searchRootSpec` rejects such a root first. It was uncovered before this branch.

## Checklist

- [x] Commits follow Conventional Commits (`type(scope): subject`, imperative, lowercase, max 72 characters)
- [x] Branch name follows the conventional branch spec (`feature/`, `bugfix/`, `chore/`, ...)
- [x] `pnpm lint`, `pnpm typecheck`, `pnpm test:coverage` and `pnpm build` pass locally
- [x] New behavior is covered by unit tests against fixtures
- [x] Live behavior is covered by e2e tests in `e2e/` where applicable
- [x] No comments and no `console` calls in changed files

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved search result handling when exact-match cards appear in different sections or use alternate data layouts.
  - Exact-match results can now be surfaced even when no standard result list is available.
  - Added clearer integrity diagnostics for section-anchor fallbacks and distinguishable search continuation failures.

- **Documentation**
  - Updated integrity diagnostics, troubleshooting guidance, and field-availability documentation.
  - Added guidance for Polish storefront search fixtures and exact-match validation scenarios.

### CODE REVIEW

File: src/features/search/search.ts [hunk 1]
```diff
@@ -76,20 +78,84 @@ export async function fetchSearchFirstPage(
...
+        `${SEARCH_CONTEXT}: exact match card resolved at sections.${scanned.sectionIndex.toString()}.${scanned.entryIndex.toString()} instead of its anchor`,
+      );
+      onIntegrityEvent?.({ context: SEARCH_CONTEXT, reason: 'section-anchor-fallback', error });
+      return scannedMatch;
+    }
+  }
+
+  const unusable = getPath(sections[0], EXACT_MATCH_MAPPINGS.card) ?? scanned?.card;
+  return reportUnusableCard(unusable, onIntegrityEvent);
+}
```
[the remark this answers is not in the document]
Comment (MrAdex77) [reply src/features/search/search.ts:153]: Confirmed and fixed in 04fe9bf. Reproduced it first: a card-shaped node missing its required fields at `sections[0][24]`, with a valid card at `sections[1][25]`, returned the list alone and emitted one `optional-section-parse`.

`scanForExactMatchCard` is now the generator `scanForExactMatchCards`, and `resolveExactMatch` takes the first yielded card that survives extraction, so the section and entry indices still name the real resolution site. Regression test: `keeps scanning past a card that cannot be extracted`, which fails against the previous implementation.

File: src/features/search/search.ts [hunk 2]
```diff
@@ -76,20 +78,84 @@ export async function fetchSearchFirstPage(
...
+    if (scannedMatch !== undefined) {
+      const error = new ParseError(
+        `${SEARCH_CONTEXT}: exact match card resolved at sections.${scanned.sectionIndex.toString()}.${scanned.entryIndex.toString()} instead of its anchor`,
+      );
+      onIntegrityEvent?.({ context: SEARCH_CONTEXT, reason: 'section-anchor-fallback', error });
+      return scannedMatch;
+    }
+  }
+
+  const unusable = getPath(sections[0], EXACT_MATCH_MAPPINGS.card) ?? scanned?.card;
```
[the remark this answers is not in the document]
Comment (MrAdex77) [reply src/features/search/search.ts]: Confirmed and fixed in the same commit, though not with the suggested swap.

The suggestion `scanned?.card ?? getPath(sections[0], ...)` reports the first scanned card, which after the generator change is no longer necessarily the one that failed. The resolution now remembers the first scanned card that failed extraction and reports that, falling back to the first anchor only when the scan yielded nothing. Same outcome in the case you described, and it stays correct once the scan continues past a failure.

Verified: with noise at `sections[0][23]` and a card missing its required fields at `sections[1][24]`, the event names four missing fields from the card paths rather than five from the noise node. Regression test: `reports the scanned card rather than an unrelated node at the first anchor`, which asserts the message does not mention `appId`.
